Run marked cells

jhamrick opened the pull request:
Fixes #677 and adds tests. They will fail though until #676 is merged as this PR relies on the assumption that `get_marked_indices` will return all marked cells including the currently selected cell, and this assumption is only introduced in #676.

minrk commented:
Looks like there's a mode-related change that's causing test failures.

ellisonbg commented on notebook/static/notebook/js/notebook.js:
```diff
@@ -1958,26 +1958,48 @@ define(function (require) {
+    /**
+     * Execute or render cell outputs and go into command mode.
+     */
+    Notebook.prototype.execute_cell = function () {
```
This name really confused me, maybe `execute_marked_cells`?

jhamrick replied:
Yeah. The reason I didn't change this name was because it is used in other locations already (the `execute_cell` function is not one I added, just modified). If you think it is going to make things clearer though I can change the name here and elsewhere.

ellisonbg commented on notebook/static/notebook/js/notebook.js:
```diff
@@ -1958,26 +1958,48 @@ define(function (require) {
      */
     Notebook.prototype.execute_cell_and_insert_below = function () {
+        // execute the marked cells, and don't insert anything
+        var indices = this.get_marked_indices();
```
Why not just call `execute_marked_cells`?

jhamrick replied:
We need the logic here of checking how many marked cells there are -- if there are multiple, then we want to execute the marked cells, otherwise we want the normal behavior. I could have it call `execute_marked_cells` (aka `execute_cell`) as well as doing the check but then it has to call `get_marked_indices` twice.

ellisonbg commented on notebook/static/notebook/js/notebook.js:
```diff
@@ -2002,10 +2024,15 @@ define(function (require) {
      */
     Notebook.prototype.execute_cell_and_select_below = function () {
+        // execute the marked cells, and don't select anything
+        var indices = this.get_marked_indices();
```
Also `execute_marked_cells` here?

jhamrick replied:
Same response as above.

ellisonbg commented on notebook/static/notebook/js/notebook.js:
```diff
@@ -2055,11 +2082,11 @@ define(function (require) {
      * @param {integer} end - index of the last cell to execute (exclusive)
      */
     Notebook.prototype.execute_cell_range = function (start, end) {
-        this.command_mode();
```
The command/edit mode stuff tends to all be in place for very specific reasons - changing them will break or change the user experience in subtle ways - often related to focus. Can you clarify why you removed this?

jhamrick replied:
I didn't truly remove this -- this method uses `execute_cells` now, which switches into command mode.

ellisonbg replied:
OK, as long as there are not code paths where that can be called twice accidentally.

jhamrick commented:
> Looks like there's a mode-related change that's causing test failures.

@minrk yeah as I mentioned, the tests will fail for this PR until #676 is merged and I can rebase on top of that.

Carreau commented on notebook/static/notebook/js/notebook.js:
```diff
@@ -1960,26 +1960,48 @@ define(function (require) {
+
+        var cell;
+        for (var i = 0; i < indices.length; i++) {
+            this.select(indices[i]);
```
do you need to select it ?

ellisonbg replied:
Hmm, yeah, I am not sure that should select each cell as it runs it. Can you explain what the logic is? I am worried that will cause a lot of focus jumping on notebooks with long cells.

jhamrick replied:
That's what the behavior was before, so I kept it. I can just make it select the last index, though, after it's done executing everything.

jhamrick commented on notebook/static/notebook/js/notebook.js:
```diff
@@ -1960,26 +1960,48 @@ define(function (require) {
-        var cell = this.get_selected_cell();
-        
-        cell.execute();
+    Notebook.prototype.execute_cells = function (indices) {
```
Yeah, probably a good idea.

jhamrick commented:
@Carreau @ellisonbg I have addressed your comments. I added back in the call to `command_mode` since I realized it actually did get called multiple times before (once in `execute_cell_range` and once each time `execute_cell` was called), so now it gets executed twice (once in `execute_cell_range` and once in `execute_cells`). I am not sure if it is really necessary to call the command mode function inside the loop, though I can have it do that, too, if necessary.

ellisonbg commented:
OK great. Let's just test the various behaviors interactively really well...

jhamrick commented:
Ok, rebased, tests should (hopefully) pass now.
